lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
-///    scattered during FIRRTL parsing to attach all the annotations associated
-///    with elements on the right components.
-///
-/// 3. Add interface ops and populate the elements.
-///
-/// Grand Central supports three "normal" element types and four "weird" element
-/// types.  The normal ones are ground types (SystemVerilog logic), vector types
-/// (SystemVerilog unpacked arrays), and nested interface types (another
-/// SystemVerilog interface).  The Chisel API provides "weird" elements that
-/// include: Boolean, Integer, String, and Double.  The SFC implementation
-/// currently drops these, but this pass emits them as commented out strings.
-struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
-  void runOnOperation() override;
-
-  // A map storing mutable information about an element in an interface.  This
-  // is keyed using a (defName, name) tuple where defname is the name of the
-  // interface and name is the name of the element.
-  typedef DenseMap<std::pair<StringRef, StringRef>, ElementInfo> InterfaceMap;
 
 private:
-  // Store a mapping of interface name to InterfaceOp.
-  llvm::StringMap<sv::InterfaceOp> interfaces;
+  SmallString<0> newNameImpl(StringRef name) {
+    // Special case the situation where there is no name collision to avoid
+    // messing with the SmallString allocation below.
+    if (internal.insert(name).second)
+      return name;
+    size_t i = 0;
+    SmallString<0> tryName;
+    do {
+      tryName = (name + "_" + Twine(i++)).str();
+    } while (!internal.insert(tryName).second);
+    return tryName;
+  }
 
-  // Discovered interfaces that need to be constructed.
-  InterfaceMap interfaceMap;
+public:
+  /// Generate a new name and add it to the namespace.
+  SmallString<0> newName(StringRef name) { return newNameImpl(name); }
 
-  // Track the order that interfaces should be emitted in.
-  SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+  /// Generate a new name and add it to the namespace.
+  SmallString<0> newName(Twine name) { return newNameImpl(name.str()); }
 };
 
-class GrandCentralVisitor : public FIRRTLVisitor<GrandCentralVisitor> {
-public:
-  GrandCentralVisitor(GrandCentralPass::InterfaceMap &interfaceMap)
-      : interfaceMap(interfaceMap) {}
+/// Stores the information content of an ExtractGrandCentralAnnotation.
+struct ExtractionInfo {
+  /// The directority where Grand Central generated collateral (modules,
+  /// interfaces, etc.) will be written.
+  StringRef directory;
 
-private:
-  /// Mutable store tracking each element in an interface.  This is indexed by a
-  /// "defName" -> "name" tuple.
-  GrandCentralPass::InterfaceMap &interfaceMap;
+  /// The name of the file where any binds will be written.  This will be placed
+  /// in the same output area as normal compilation output, e.g., output
+  /// Verilog.  This has no relation to the `directory` member.
+  StringRef bindFilename;
```

Comment:
It looks like it might be possible to store `directory` and `bindFilename` as StringAttr instead of StringRefs. Looks like they just go through the `StringAttr -> StringRef -> StringAttr` cycle.

Reply:
Good thought. Switching to this.